---
title: "GSE101638"
author: "Kaitlyn O'Shea"
output: html_document
---

**THIS SCRIPT HAS BEEN ARCHIVED DUE TO THE AVAILABILITY OF TPM DATA FORMAT ONLY, LIMITING THE ABILITY TO LOOK FOR RELATIONSHIPS ACROSS SAMPLES DUE TO WITHIN SAME NORMALIZATION - CAN REVISIT IF COUNT MATRIX BECOMES PUBLICALLY AVAILABLE**

```{r setup}
knitr::opts_chunk$set(echo = TRUE,
                      eval = TRUE)
```

```{r install_packages, eval = FALSE}
## require BioManager
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GEOquery") 
```

```{r libraries, message=FALSE, warning=FALSE}
## libraries needed
library("tidyverse")
library("GEOquery")
library("here")
```

## Get GSE101638 from GEO 
```{r Geo_download, message=FALSE, warning=FALSE}
## Download dataset from GEO, save downloaded files in the same folder as this .Rmd file.
dir.create(file.path("raw","gse101638"), recursive = TRUE)
gse101638 <- getGEO("gse101638",  # sample accession 
                   destdir = file.path("raw","gse101638"),
                   GSEMatrix = TRUE,
                   AnnotGPL = TRUE)
```

## Clinical Data from Dataset
```{r ClinicalDataExtract}
## index into data list and extract clinical data
gse101638_cd <- gse101638[["GSE101638_series_matrix.txt.gz"]] %>%   # index into list 
                  phenoData() %>%                                   # select phenoData
                  pData() %>%                                       # apply pData
                  as_tibble() %>%                                   # tidy friendly
                  select("geo_accession" = geo_accession,
                         "sample_id" = `tissue id:ch1`,
                         tissue = "meningioma"
                         ) 

gse101638_pheno <- gse101638_cd %>%         # rename and adjust variables
                    #   mutate(grade = factor(grade,
                    #                         levels = c("1", "2", "3"), 
                    #                         labels = c("I", "II", "III")
                    #                         ),
                    #          
                    # ) %>%
                      mutate_if(is.character, as.factor)
```

## rnaseq Data from Dataset
```{r ExpressionDataExtract}
## download and save the TPM file 
gse101638_exprs <-          # tidy friendly

## match transcripts to hg19 transcript names    
gse101638_feature <-    # tidy friendly

```

## Transform rnaseq Data
``` {r ExpressionDataTransform}
# function to find standard deviation of each probe across all tumors
rowsd <- function(x) {
                      sqrt(
                        rowSums((x - rowMeans(x))^2)/
                          (dim(x)[2] - 1)
                        )
}

gse101638_rnaseq <- bind_cols(gse101638_feature["UCSC_RefGene_Name"], gse101638_exprs) %>% # add gene name to data
               rename("gene" = "UCSC_RefGene_Name") %>%                       # rename name variable as 'gene' 
               filter(gene != "") %>%                                         # keep only data for named genes
               mutate(sd = rowsd(.[,-1])) %>%                                 # calculate sd for each probe, index out gene name
               group_by(gene) %>%                                             # group probes by gene
               filter(sd == max(sd)) %>%                                      # keep the probe with the max variation for each gene
               ungroup() %>%                                                  # ungroup by genes 
               select(-sd) %>%                                                # select all but sd column
               pivot_longer(-gene,                                            # do not group by gene 
                            names_to = "geo_accession",                       
                            values_to = "value") %>%                          
               mutate("mvalue" = log2(value/(1 - value))) %>%                 # log2 transform B / (1-B) expression for variance stabilization 
               group_by(gene) %>%                                             # standardize for cross data set comparison
               mutate("standardized" = (mvalue - mean(mvalue))/               # center at mean for gene 
                                         sd(mvalue),                          # scale by sd for gene
                       ) %>%
                ungroup() %>%                                                 # ungroup by genes 
                select(-value, -mvalue) %>%                                           # select all but sd column
                pivot_wider(names_from = "gene",                              # final format genes as rows
                            values_from = "standardized")                     # final format expression across accession as columns    
```

## Dataset for Analysis
```{r FinalDataset}
## combine phenotype and expression data for workable data
gse101638_data <- inner_join(gse101638_pheno, gse101638_rnaseq, 
                            by = "geo_accession")   # match by accession

## create metatag for dataset
gse101638_info <-  
                list(
                     name = "Harmanci_a",               # author
                     accession = "GSE101638",         # dataset accession 
                     # the following are logical to determine clinical data available for dataset
                     geo_accession = "geo_accession" %in% colnames(gse101638_pheno),
                      sample_id = "sample_id" %in% colnames(gse101638_pheno),
                      age = "age" %in% colnames(gse101638_pheno),
                      sex = "sex" %in% colnames(gse101638_pheno),
                      grade = "grade" %in% colnames(gse101638_pheno),
                      histology = "histology" %in% colnames(gse101638_pheno),
                      type = "type" %in% colnames(gse101638_pheno),
                      location = "location" %in% colnames(gse101638_pheno),
                      mib1_index = "mib1_index" %in% colnames(gse101638_pheno),
                      recurrence = "recurrence" %in% colnames(gse101638_pheno),
                      recurrence_yrs = "recurrence_yrs" %in% colnames(gse101638_pheno),
                      recurrence_freq = "recurrence_freq" %in% colnames(gse101638_pheno),
                      followup_days = "followup_days" %in% colnames(gse101638_pheno),
                      survival_yrs = "survival_yrs" %in% colnames(gse101638_pheno),
                      vital_indicator = "vital_indicator" %in% colnames(gse101638_pheno),
                      tissue = "tissue" %in% colnames(gse101638_pheno),
                      progression = "progression" %in% colnames(gse101638_pheno),
                      radiation = "radiation" %in% colnames(gse101638_pheno),
                      rna_integrity ="rna_integrity" %in% colnames(gse101638_pheno),

                     sample_numbers = nrow(gse101638_data) # number of samples 
                     )

## create dataset as list 
Harmanci_a_2013 <- list(
                dataset_info = gse101638_info,             # metatag
                clinical_data = gse101638_pheno,           # clinical data
                expression_data = gse101638_rnaseq,           # expression data
                data = gse101638_data #,                     # combined, workable data
             #   full_GEO = gse101638                       # full original download from GEO
)
```

```{r}
save(Harmanci_a_2013, file = here("datasets/rnaseq","GSE101638_Harmanci_A.Rda")) 
```